sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer.py

```diff
@@ -0,0 +1,221 @@
+                    )
+                )
+        if flush:
+            self.flush(timeout=timeout, raise_error=False)
+        else:
+            if self._cur_buffered_len:
+                _LOGGER.warning(
+                    "Shutting down Partition {}."
+                    " There are still {} events in the buffer which will got lost".format(
+                        self.partition_id,
+                        self._cur_buffered_len
+                    )
+                )
+        self._producer.close()
+
+    def put_events(self, events, timeout=None):
+        # Put single event or EventDataBatch into the queue.
+        # This method would raise OperationTimeout if the queue does not have enough space for the input and
+        # flush cannot finish in timeout.
+        with self._not_full:
+            try:
+                new_events_len = len(events)
+            except TypeError:
+                new_events_len = 1
+            timeout_time = time.time() + timeout if timeout else None
+
+            if self._max_buffer_len - self._cur_buffered_len < new_events_len:
+                _LOGGER.info(
+                    "Partition {} does not have enough room for coming {} events."
+                    "Flush first.".format(
+                        self.partition_id, new_events_len
+                    )
+                )
+                # flush the buffer
+                self.flush(timeout=timeout)
+
+            if timeout_time and time.time() > timeout_time:
+                raise OperationTimeoutError("Failed to enqueue events into buffer due to timeout.")
+
+            try:
+                # add single event into current batch
+                self._cur_batch.add(events)
+            except AttributeError:
+                # if the input events is a EventDataBatch, put the whole into the buffer
+                self._buffered_queue.put(events)
+                # create a new batch for incoming events
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                # put the new batch into the buffer
+                self._buffered_queue.put(self._cur_batch)
+            except ValueError:
+                # add single event exceeds the cur batch size, create new batch
+                self._cur_batch = EventDataBatch(self._max_message_size_on_link)
+                self._buffered_queue.put(self._cur_batch)
+                self._cur_batch.add(events)
+            self._cur_buffered_len += new_events_len
+            # notify the max_wait_time worker
+            self._not_empty.notify()
+
+    def enhanced_callback_decorator(self, callback):
+        def wrapper_callback(*args, **kwargs):
+            try:
+                callback(*args, **kwargs)
+            except Exception as exc:
+                _LOGGER.warning("callback exception", callback.__name__, exc, self.partition_id)
+
+        return wrapper_callback
+
+    def sent_callback(self, events, future: Future):
```

Comment:
Where is this used?

Reply:
this is originally used for the concurrent sending feature, but I'm removing it for now as we've decided to pull that out of the scope